src/Bicep.Core/Emit/TemplateWriter.cs

```diff
@@ -85,34 +87,50 @@ private static string GetSchema(ResourceScope targetScope)
 
             return "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#";
         }
+
         private readonly EmitterContext context;
         private readonly EmitterSettings settings;
+        private readonly RawSourceMap? rawSourceMap;
+
+        public SourceMap? SourceMap; // ARM JSON line => (Bicep File, Bicep Line)
 
         public TemplateWriter(SemanticModel semanticModel, EmitterSettings settings)
         {
             this.context = new EmitterContext(semanticModel, settings);
             this.settings = settings;
+            this.rawSourceMap = (this.context.Settings.EnableSourceMapping)
+                ? new(new List<RawSourceMapFileEntry>())
+                : null;
+            this.SourceMap = null;
         }
 
         public void Write(JsonTextWriter writer)
         {
-            // Template is used for calcualting template hash, template jtoken is used for writing to file.
+            // Template is used for calculating template hash, template JToken is used for writing to file.
             var (template, templateJToken) = GenerateTemplateWithoutHash();
```

Comment:
Once again, just thinking about avoiding state and trying to have a cleaner separation of concerns - can we:
1. Add an interface `ISourceAwareJsonWriter` for `PositionTrackingJsonTextWriter`, and add any necessary methods.
1. Remove `rawSourceMap` & `SourceMap` from this class (and have the `PositionTrackingJsonWriter` be responsible for managing their lifecycles).
1. Have `Write` take the `ISourceAwareJsonWriter` interface as a parameter.
1. Make it the responsibility of `TemplateEmitter` to instantiate and pass in the `PositionTrackingJsonWriter` with source mapping enabled/disabled as per the settings.